# Use Customer_Segmentation Notebook Template

In [1]:
import numpy as np # linear algebra
import pandas as pd

import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report, confusion_matrix

import warnings
warnings.filterwarnings('ignore')

Matplotlib created a temporary cache directory at /tmp/matplotlib-cjiq6pu5 because the default path (/home/mosaic-ai/.cache/matplotlib) is not a writable directory; it is highly recommended to set the MPLCONFIGDIR environment variable to a writable directory, in particular to speed up the import of Matplotlib and to better support multiprocessing.


# Read the intermeditary output of K-Means clustering Model Building Notebooks for model training and registration

In [2]:
df = pd.read_csv('/data/Output/kmeans_clustered_customer_Data.csv')

In [3]:
df.shape

(122882, 29)

In [4]:
df.tail()

CUSTOMER_ID        DATE  AGE     SEX  \
122877  3a7ce3d5-0f4b-4575-badc-3553f5440c92  2023-05-01   32    MALE   
122878  e2a8f308-57eb-4f55-bc13-399378f246d3  2022-12-01   27    MALE   
122879  599c0e61-9bbc-4e14-bbfd-b07f5a398669  2022-12-01   33  FEMALE   
122880  77298902-f899-4c3e-9104-06d1736fb089  2023-06-01   33  FEMALE   
122881  cd75294e-c31f-41b3-b78c-1d3e7322ddcf  2022-12-01   44  FEMALE   

       EDUCATION_LEVEL EMPLOYMENT_STATUS      HOBBIES MARITAL_STATUS  \
122877     High School        Unemployed      reading         Single   
122878             PhD        Unemployed      camping         Single   
122879     High School        Unemployed  video-games         Single   
122880         College        Unemployed   basketball         Single   
122881     High School          Employed     exercise        Married   

            DEPENDENTS REGISTRATION_DATE  ... INSTALLMENTS_PURCHASES  \
122877   One Dependent        2020-03-12  ...              339.07230   
122878   One Dependent        2009-05-12  ...              482.58685   
122879   One Dependent        2007-02-12  ...                0.00000   
122880   One Dependent        2001-02-12  ...                0.00000   
122881  Two Dependents        2001-02-12  ...                0.00000   

        PURCHASES_FREQUENCY ONEOFF_PURCHASES_FREQUENCY  \
122877            Very high                        Low   
122878            Very high                   Very low   
122879             Very low                   Very low   
122880             Very low                   Very low   
122881             Very low                   Very low   

        PURCHASES_INSTALLMENTS_FREQUENCY  PURCHASES_TRX     PAYMENTS  \
122877                         Very high          21.20   273.352691   
122878                         Very high          25.92   474.722440   
122879                          Very low           0.00   568.542708   
122880                          Very low           0.00   517.588408   
122881                          Very low           0.00  1029.696570   

        MINIMUM_PAYMENTS  PRC_FULL_PAYMENT  Cluster           Cluster_Type  
122877        242.288921               0.0        0  Conservative Spenders  
122878        165.534652               0.0        0  Conservative Spenders  
122879        549.300475               0.0        0  Conservative Spenders  
122880        544.214360               0.0        0  Conservative Spenders  
122881       1552.899035               0.0        2     Cash Advance Users  

[5 rows x 29 columns]

In [5]:
to_drop = ['CUSTOMER_ID','DATE','AGE','SEX','EDUCATION_LEVEL','EMPLOYMENT_STATUS','HOBBIES','MARITAL_STATUS','BANK_ACCOUNT_TYPE',
 'DEPENDENTS','REGISTRATION_DATE','PURCHASES_FREQUENCY','ONEOFF_PURCHASES_FREQUENCY','PURCHASES_INSTALLMENTS_FREQUENCY','BALANCE_FREQUENCY','Cluster_Type']

In [6]:
df.drop(to_drop, axis= 1, inplace= True)

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 122882 entries, 0 to 122881
Data columns (total 13 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   BALANCE                 122882 non-null  float64
 1   CREDIT_LIMIT            122882 non-null  float64
 2   CASH_ADVANCE            122882 non-null  float64
 3   CASH_ADVANCE_FREQUENCY  122882 non-null  float64
 4   CASH_ADVANCE_TRX        122882 non-null  float64
 5   PURCHASES               122882 non-null  float64
 6   ONEOFF_PURCHASES        122882 non-null  float64
 7   INSTALLMENTS_PURCHASES  122882 non-null  float64
 8   PURCHASES_TRX           122882 non-null  float64
 9   PAYMENTS                122882 non-null  float64
 10  MINIMUM_PAYMENTS        122882 non-null  float64
 11  PRC_FULL_PAYMENT        122882 non-null  float64
 12  Cluster                 122882 non-null  int64  
dtypes: float64(12), int64(1)
memory usage: 12.2 MB


In [9]:
X = df.drop(['Cluster'],axis = 1 )
y = df[['Cluster']]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3,random_state = 101)

In [10]:
X_train.shape, X_test.shape

((86017, 12), (36865, 12))

# Decision Tree Classifier

In [11]:
#Training our algorithm
clf = DecisionTreeClassifier()
clf.fit(X_train, y_train)
clf_prediction = clf.predict(X_test)

In [12]:
#Checking our prediction accuracy score
print(confusion_matrix(y_test, clf_prediction))

[[25607     0   135   122]
 [    0   311     0    10]
 [  111     0  5939    13]
 [  153     7    17  4440]]


In [13]:
print(classification_report(y_test, clf_prediction))

              precision    recall  f1-score   support

           0       0.99      0.99      0.99     25864
           1       0.98      0.97      0.97       321
           2       0.98      0.98      0.98      6063
           3       0.97      0.96      0.97      4617

    accuracy                           0.98     36865
   macro avg       0.98      0.98      0.98     36865
weighted avg       0.98      0.98      0.98     36865



## We achieve 98% accuracy at predicting potiential new customers behavior